### PR TITLE
amount:  create and return integer values

### PR DESCRIPTION
**Pull request by Kunde21**
Enable handling of 64-bit integer and big integer amount values in
minor units.  These types are seen in financial and banking APIs to
avoid the possibility of invalid amounts when dealing with multiple
currencies that have varying decimal handling.

Related to #5 (implementation can be done separately with these changes in place)

### CODE REVIEW

File: amount.go [hunk 1]
```diff
@@ -71,10 +72,26 @@ func NewAmount(n, currencyCode string) (Amount, error) {
...
+	if currencyCode == "" || !IsValid(currencyCode) {
+		return Amount{}, InvalidCurrencyCodeError{"NewAmountFromBigInt", currencyCode}
+	}
+	d, _ := GetDigits(currencyCode)
+	return Amount{apd.NewWithBigInt(amt, -int32(d)), currencyCode}, nil
+}
+
+// NewAmount creates a new Amount from an int64 and a currency code.
+func NewAmountFromInt64(amt int64, currencyCode string) (Amount, error) {
+	return NewAmountFromInt64(amt, currencyCode)
```
Comment (bojanz) [inline amount.go:92]: What happened here? The func is calling itself?

File: amount.go [hunk 2]
```diff
@@ -112,7 +143,6 @@ func (a Amount) Convert(currencyCode, rate string) (Amount, error) {
 	}
 	ctx := apd.BaseContext.WithPrecision(16)
 	ctx.Mul(result, a.number, result)
-
```
Comment (bojanz) [inline amount.go:115]: Let's not revert my code formatting choices. I like having a newline before the return.

**Comment by Kunde21** [comment]
Actually a bit surprised by how badly I borked that rebase.  :grimacing: 

Fixed and squashed, so it's only the changes I intended to make.

**Comment by bojanz** [comment]
Thanks, that looks much better.

Can we add some test coverage for the new methods?

File: amount.go [hunk 3]
```diff
@@ -75,6 +76,24 @@ func NewAmount(n, currencyCode string) (Amount, error) {
 	return Amount{number, currencyCode}, nil
 }
 
+// NewAmountFromBigInt creates a new Amount from a big integer and a currency code.
+func NewAmountFromBigInt(amt *big.Int, currencyCode string) (Amount, error) {
```
Comment (bojanz) [inline amount.go:80]: Let's use "n" instead of "amt" to match NewAmount().

File: amount.go [hunk 4]
```diff
@@ -75,6 +76,24 @@ func NewAmount(n, currencyCode string) (Amount, error) {
...
+		return Amount{}, InvalidCurrencyCodeError{"NewAmountFromBigInt", currencyCode}
+	}
+	d, _ := GetDigits(currencyCode)
+
+	return Amount{apd.NewWithBigInt(amt, -int32(d)), currencyCode}, nil
+}
+
+// NewAmount creates a new Amount from an int64 and a currency code.
+func NewAmountFromInt64(amt int64, currencyCode string) (Amount, error) {
+	return NewAmountFromBigInt(big.NewInt(amt), currencyCode)
```
Comment (bojanz) [inline amount.go:94]: I like this, it matches what apd.New() does under the hood (wrap the int64 with a big.Int), and reduces boilerplate.

File: amount.go [hunk 5]
```diff
@@ -181,13 +217,28 @@ func (a Amount) RoundTo(digits uint8, mode RoundingMode) Amount {
...
 	result := apd.New(0, 0)
-	ctx := apd.BaseContext.WithPrecision(16)
+	ctx := contextPrecision(a.number.NumDigits())
 	ctx.Rounding = extModes[mode]
 	ctx.Quantize(result, a.number, -int32(digits))
 
 	return Amount{result, a.currencyCode}
 }
 
+func contextPrecision(digits ...int64) *apd.Context {
```
Comment (bojanz) [inline amount.go:227]: What is the reason behind introducing this logic? In which cases was the hardcoded 16 not enough?

Comment (Kunde21) [reply amount.go:227]: Testing with `math.MaxInt64` failed, returning `0` from a call to `Round()`

File: amount.go [hunk 6]
```diff
@@ -116,25 +117,25 @@ func (a Amount) String() string {
...
 }
 
-// Int64 returns the integer value of a in minor units.
-// Returns an error if value can't be expressed as a 64-bit integer.
+// Int64 returns a in minor units, as an int64.
+// If a cannot be represented in an int64, an error is returned.
 func (a Amount) Int64() (int64, error) {
-	n := *a.Round().number
+	n := a.Round().number
 	n.Exponent = 0
```
Comment (Kunde21) [inline amount.go]: This may change the underlying number.  I copied the value to avoid that.

Comment (bojanz) [reply amount.go]: Whoops, that was unintentional.

**Comment by bojanz** [comment]
I finally have time to wrap this PR up. I'll continue pushing tweaks over the next day or two.

**Comment by Kunde21** [comment]
Quick follow-up on this PR.  Looking to migrate a few projects off of my fork, to clean things up a bit.

**Comment by bojanz** [comment]
Wrapped up and committed the API parts of this PR in e18a43233c0ec25e3ad0d130e580d56d6d031158. Sorry it took this long.

Opened #10 to deal with contextPrecision, as I think we want expanded test coverage in multiple places. Keeping this PR open until I have the precision changes up in a new PR.

**Comment by bojanz** [comment]
Opened https://github.com/bojanz/currency/pull/11 for the other half, let's wrap it up there.

**Pull request closed without merging**